File: p2/testing.cpp
#include <mutex>
#include <condition_variable>
#include <functional>
#include <thread>
#include <iostream>
#include <random>

void chaos();

class FooBar {
private:
    int n;
    std::mutex m;
    std::condition_variable cv;
    bool printingBar;

public:
    FooBar(int n) {
        this->n = n;
        printingBar = false;
    }

    void oneStep(std::function<void()> printFunc, bool isBar) {
        {
            chaos();
            std::unique_lock<std::mutex> lk(m);
            chaos();
            cv.wait(lk, [this, isBar](){return isBar == printingBar;});
            chaos();
        } 
        chaos();
        printFunc();
        chaos();
        {
            chaos();
            std::unique_lock<std::mutex> lk(m);
            chaos();
            printingBar = !isBar;
            chaos();
        }
        chaos();
        cv.notify_one();
    }

    void foo(std::function<void()> printFoo) {      
        for (int i = 0; i < n; i++) {
            oneStep(printFoo, false);
        }
    }

    void bar(std::function<void()> printBar) {
        for (int i = 0; i < n; i++) {
            oneStep(printBar, true);
        }
    }
};

void chaos() {
    static std::mt19937 rng(std::random_device{}());
    std::uniform_int_distribution<std::chrono::milliseconds::rep> dist(0, 5);
    std::this_thread::sleep_for(std::chrono::milliseconds(dist(rng)));
}

void printFoo() {
    std::string s = "foo";
    for (char c : s) {
        std::cout << c;
        std::cout.flush();
        chaos();
    }
}

void printBar() {
    std::string s = "bar";
    for (char c : s) {
        std::cout << c;
        std::cout.flush();
        chaos();
    }
}

int main() {
    const int repeat = 10000;
    FooBar foobar(repeat);

    std::thread t1(&FooBar::foo, &foobar, printFoo);
    std::thread t2(&FooBar::bar, &foobar, printBar);

    t1.join();
    t2.join();

    return 0;
}

File: p2/testing.java
package p2;

import java.util.concurrent.ThreadLocalRandom;

class FooBar {
    private int n;
    private boolean printingBar;

    public FooBar(int n) {
        this.n = n;
        this.printingBar = false;
    }

    private void oneStep(Runnable printFunc, boolean isBar) throws InterruptedException {
        chaos();
        synchronized(this){
            chaos();
            while(isBar != printingBar){
                chaos();
                wait();
                chaos();
            }
            chaos();
        } 
        chaos();
        printFunc.run();
        chaos();
        synchronized(this){
            chaos();
            printingBar = !isBar;
            chaos();
            notify();
            chaos();
        }
    }

    public void foo(Runnable printFoo) throws InterruptedException { 
        for (int i = 0; i < n; i++) {       
            oneStep(printFoo, false);
        }
    }

    public void bar(Runnable printBar) throws InterruptedException {
        for (int i = 0; i < n; i++) {       
            oneStep(printBar, true);
        }
    }

    private static void chaos() {
        try {
            Thread.sleep(ThreadLocalRandom.current().nextInt(6));
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public static void printFoo() {
        String s = "foo";
        for (char c : s.toCharArray()) {
            System.out.print(c);
            chaos();
        }
    }

    public static void printBar() {
        String s = "bar";
        for (char c : s.toCharArray()) {
            System.out.print(c);
            chaos();
        }
    }

    public static void main(String[] args) {
        final int repeat = 10000;
        FooBar foobar = new FooBar(repeat);

        Thread t1 = new Thread(() -> {
            try {
                foobar.foo(FooBar::printFoo);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        });

        Thread t2 = new Thread(() -> {
            try {
                foobar.bar(FooBar::printBar);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        });

        t1.start();
        t2.start();

        try {
            t1.join();
            t2.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}